### PR TITLE
[WIP] Welcome wizard

### PR DESCRIPTION
**Pull request by bumi**
The welcome wizard helps the user to configure the account and get started. 

See miro and figma boards for flow and design details

[ref #3]

### CODE REVIEW

**Comment by bumi** [comment]
@reznord do you have a suggestion on what's an easy way to keep state between two components? e.g. in this case between two router steps (LndSetup and LndConfirm for example?)

**Comment by chris-aeviator** [comment]
@bumi https://redux.js.org/

**Comment by bumi** [comment]
@chris-aeviator thanks, do you recommend we should introduce redux?

**Comment by chris-aeviator** [comment]
It is possible to go completely without a store when refactoring so that the extension's main entrypoint (e.g. a not yet existing `app.js`) holds all state and passes it via props (https://reactjs.org/docs/components-and-props.html). You can also pass setter functions like `const myFancyProp, setMyFancyProp = useState()`, defined by state hooks (https://reactjs.org/docs/hooks-state.html)...I believe this works the same with class based state hooks as used in this extension.

I think it boils down to the question if we will further down the road have more state to manage. Nuxt/Vue uses a system like redux natively and it works pretty good. Redux is the go to solution for many in react.

**Comment by bumi** [comment]
hard for me to decide because I lack the react experience. 

the use case that I have here is the following:
1. screen: allow the user to set a password  
2. screen: ask for node config
3. then encrypt the node config with the password given in 1. screen. and store the encrypted config

**Comment by chris-aeviator** [comment]
I'll have some free time in the early morning tomorrow and will take a look, it sounds to me that with some refactoring we can go without.

File: source/Welcome/Welcome.js [hunk 1]
```diff
@@ -1,45 +1,42 @@
...
+    this.accountsStore.reset();
   }
 
   render() {
     return (
-      <div>
-        <p>Hallo</p>
-        <span
-          onClick={() => {
-            this.reset();
```
Comment (chris-aeviator) [inline source/Welcome/Welcome.js:37]: reset functionality has gone nirvana - any replacement?

**Comment by chris-aeviator** [comment]
@bumi no refactoring was necessary. I showcased how to pass data (that's hold by a central state in https://github.com/bumi/lightning-browser-extension/blob/0135057d9e5dc24346803a6bc105f7c0702afb09/source/Welcome/Welcome.js ) from the _Setup_ step to the _Confirm_ step of the welcome wizard. If you agree we can discuss what data really should be passed between which screens and implement the single screens.

**Comment by reznord** [comment]
> @reznord do you have a suggestion on what's an easy way to keep state between two components? e.g. in this case between two router steps (LndSetup and LndConfirm for example?)

Sorry for the late reply, but since the content passing is between two different routes of the same page, it is better to go with redux as @chris-aeviator suggested earlier.

**Comment by bumi** [comment]
@reznord why do you suggest redux instead of @chris-aeviator's simpler solution to pass state to the central Welcome.js?

**Comment by reznord** [comment]
The reason why I suggested to use redux is to avoid prop-drilling through the components. I would totally agree with just passing it as a prop to the component provided we write the entire view logic for the page in one single component. But that makes readability a lot harder.

So, to avoid issues like this, there are two options:
- Pass the prop to the page as a `prop` and then on the page level add it on the context provider where you can share the required prop across all components by avoiding prop drilling. The problem with this approach is that you will have a lot of manual boilerplate to handle for each and every page. 
- Do the store logic manipulations on the redux store and then you can access the data from the store directly in any component. This advantages of this particular approach would be less boilerplate code (since you just need to wrap the redux  provider only on the top level where you render the app) and there won't be any extra logic on each screen to handle the prop drilling etc.

This is the reason why I suggested to go with redux.

**Comment by chris-aeviator** [comment]
I agree with @reznord 's view in partial. Since we're just passing a minimal amount of information here it will not grow anymore unreadable as of now though.

Bumi defined 

> 
1.  screen: allow the user to set a password
2.    screen: ask for node config
3.   then encrypt the node config with the password given in 1. screen. and store the encrypted config

and that's it. I'd say if it comes to a point where we need to work with more state logic (and I don't see it at the moment) we could re-discuss redux but for now my view is to save our resources and concentrate on coding the screens?!

We can further reduce the props passed from e.g.

```
      <LndConfirm
                lndName={this.state.lndName}
                lndMacaroon={this.state.lndMacaroon}
                lndUrl={this.state.lndUrl}
              />
```

to 

```
      <LndConfirm
                lndConfig={this.state.lndConfig}
              />
```

**Comment by chris-aeviator** [comment]
btw. for better explanation to @bumi - phraseology like "prop drilling" boils down to paradigms around communities.

The react community (not react itself) discourages the concept and calls it prop drilling, emberJS (and I believe also vue) for example encourages it and calls it "data down – actions up".

 I think it more boils down to a "believe". Following that we'd also need to discuss to rewrite everything to functional components instead of class based components, since this is where the community seems to be heading.

**Comment by bumi** [comment]
thanks for that explanation! puh... :)  
Generally I just think it is important to follow best practices and standards as much as possible. 

but that means we go for now with passing in the props and if it becomes too much and complicated we move to redux? - and that should be rather simple to do then?

**Comment by chris-aeviator** [comment]
the only thing I find in the react docs about it is

"A component may choose to pass its state down as props to its child components:" (https://reactjs.org/docs/state-and-lifecycle.html#the-data-flows-down)

**Comment by chris-aeviator** [comment]
will continue with the welcome wizard, open to any refactoring shortly thereafter.

**Comment by bumi** [comment]
closing in favor of #44

**Pull request closed without merging**